### ISSUE TITLE
ResNet model failing to load in C# project

### ISSUE THREAD

**Issue by achhibbar** [Microsoft/CNTK#999]
I have trained AlexNet and ResNet_18 models by adapting the example .cntk and .ndl files provided. The major difference being output dimensions being 2 (rather than 10). Both were trained on the GPU.
I am able to load my trained AlexNet model into the C# Eval project and test it on new images. However when I do the same with ResNet, I get an error:
"There is no disk in the drive. Please insert a disk into drive _:."

There is also a warning about the epsilon value for BN, but I don't think this is stopping the model from loading.
An exception is also thrown within Visual Studio: "This code is compiled with CPUONLY macro".

I also tried using the pre-trained ResNet_18.model from https://www.cntk.ai, and it worked just fine with no errors. I noticed that this pre-trained model was also slightly larger in terms of file size compared to the model that I had trained.

**Comment by zhouwangzw** [comment]
The ResNet in the example is using Batch Normalization, which is not supported on CPU, but the C# eval published in the Nuget package supports CPU-only mode (https://github.com/Microsoft/CNTK/wiki/NuGet-Package#current-release). The pre-trianed ResNet_18.model does not use BN, so it works fine.  As work around, if you could use EvalWrapper.dll and its dependencies from CNTK GPU binaries (instead of the Nuget package), you should be able to use your trained Resnet model in your program. The https://github.com/Microsoft/CNTK/blob/master/Tests/EndToEndTests/EvalClientTests/CSEvalClientTest/CSEvalClientTest.csproj shows you how to reference to EvalWrapper.dll in a C# project. Please note that you also need all dependency dlls. For long term, we are currently working on Eval C# V2 API which will support both CPU and GPU.

**Comment by YGlamazdin** [comment]
Hello. I made an example to FastRCNN in Python. Everything works normally through SNTK.exe. But when you start training model through EvalWrapper.dll with # model does not work. The model was successfully loaded, the data is processed, but the output is not looking for classroom teaching. I think the reason is the different functions of the image processing. I use EhtractCHW. Prompt please, in what could be the cause of the error?

**Comment by zhouwangzw** [comment]
@YGlamazdin, by 
> "when you start **training** model through EvalWrapper.dll with # model does not work. ", 
do you mean to train the model, or evaluate a pre-trained model? The EvalWrapper.dll can only be used to evaluate a model, but not training. With CNTK V2 API (available in Python and C++), you are able to train and evaluate model from your application. (The C# interface for CNTK V2 API will be available later). 

Regard image processing, the EvalWrapper.dll uses .Net Bitmap and its extensions (in CNTKBitmapExtensions), while the image reader and cntk.exe is using OpenCV libraries. And for FastRCNN, you probably also need some additional preprocessing by yourself. Please see  #977 for more details.

**Comment by YGlamazdin** [comment]
> also need some additional preprocessing by yourself
Is it possible to train the model immediately, so as not to require any change? Loaded with a model example from #977 of successfully working through EvalWrapper. But my, I teach myself, no. Prompt, what parameters is better to put in ImageDeserializer ?

**Comment by zhouwangzw** [comment]
@YGlamazdin The addtional preprossing is only needed when you use EvalWrapper.dll to evaluate a model, in order to make sure that input image is resized to 1000x1000, and its aspect ration is preseved by padding, and the ROIs are also adapated. However, this step is not needed for training a model, because ImageDeserilizer will take care of preprocessing based on settings in the configuratio file. The examples in https://github.com/Microsoft/CNTK/tree/master/Examples/Image/Detection/FastRCNN show you how to set train parameters for FastRCNN.

**Comment by achhibbar** [comment]
@zhouwangzw Thanks, I will try out what you have suggested.

**Comment by achhibbar** [comment]
I am attempting to open the CSEvalClientTest.csproj on my machine, and it's producing an error in Visual Studio. "The expression  (path to csproj file).Substring(88) cannot be evaluated. Startindex cannot be larger than the length of the string"

Is this file mean't to be opened on another machine as is? Or is the idea that I use it to create my project, using it as a guide?

Once this is running, I was planning to copy all of the DLL's from the CNTK release folder into this project's release folder.

**Comment by zhouwangzw** [comment]
The CSEvalClientTest.csproj is a part of cntk.sln. Can you open the project from cntk.sln? If you just want to use CSEvalClient only, the [CSEvalClient.csproj](https://github.com/Microsoft/CNTK/blob/master/Examples/Evaluation/CSEvalClient/CSEvalClient.csproj) in the Examples/Evaluation/CSEvalClient is the better choice, as this one uses Nuget package without any further configuration. You can see [README.md](https://github.com/Microsoft/CNTK/blob/master/Examples/Evaluation/README.md) for more information.

**Comment by achhibbar** [comment]
This is now working for me without crashing - thanks for your help.
However my trained ResNet model is outputting exactly the same results regardless of what image I use as input. In addition to looking at the max value of the output (like in the provided example), I am checking the actual values of the outputs also. The values never seem to change. This is not happening with my AlexNet model.

**Comment by zhouwangzw** [comment]
Please see  #1017 for more details.
